src/carbon_sendgrid_adapter.cr

```diff
@@ -2,15 +2,37 @@ require "http"
 require "json"
 require "carbon"
 
+
 class Carbon::SendGridAdapter < Carbon::Adapter
-  private getter api_key : String
+  getter authentication_strategy : AuthenticationStrategy
   private getter? sandbox : Bool
 
-  def initialize(@api_key, @sandbox = false)
+  def initialize(api_key : String, @sandbox = false)
+    @authentication_strategy = AuthenticationStrategy.new(api_key)
+  end
+
+  def initialize(username : String, password : String, @sandbox = false)
+    @authentication_strategy = AuthenticationStrategy.new(username, password)
   end
 
   def deliver_now(email : Carbon::Email)
-    Carbon::SendGridAdapter::Email.new(email, api_key, sandbox?).deliver
+    Carbon::SendGridAdapter::Email.new(email, authentication_strategy, sandbox?).deliver
+  end
+
+  class AuthenticationStrategy
+    private getter api_key : String?
+    private getter username : String?
+    private getter password : String?
+
+    def initialize(@username, @password)
+    end
+
+    def initialize(@api_key)
+    end
+    
+    def authenticate_with_basic_auth?
+      api_key.nil?
```

Comment:
If no error was thrown during the construction of the adapter then a nil api_key means that _USER and _PASSWORD were present and therefore a basic auth is in order